### ISSUE TITLE
Prevent Hazelcast client from shutting down on connection problems

### ISSUE THREAD

**Issue by entbugger** [hazelcast/hazelcast#19429]
**The problem we are trying to solve**
We have a web application (Spring) that lazy connects to Hazelcast. The client is initialized on first HTTP request and is configured via XML (with the `<hazelcast-client>` element) through the XML configuration of a third-party library.

When Hazelcast server stays unreachable for a longer time (e.g because of network) the client will try on the first invocation for `connection-attempt-limit` times and then shut down for always. When the server gets up again no requests will be ever made because the client is in the SHUTDOWN state.

**Desired behavior**
We need the client to try reconnecting **for every invocation** for a limited amount of times even when the server is not reachable for longer instead of shutting down.

This is an example of what we want when the server is not reachable:
    At 11:00:00 client.invoke() -> client should try X times (for a limited amount of time) and fail (without client shutdown)
    At 11:02:00 client.invoke() -> client should try X times (for a limited amount of time) and fail (without client shutdown)
    At 16:00:00 client.invoke() -> client should try X times (for a limited amount of time) and fail (without client shutdown)

This is similar to Tomcat JDBC Connection Pool when configured with the `testOnBorrow=true` feature.

**Alternatives we have considered**
There are some parameters to tweak the client, but no combination of these seems to be enough. E.g:

1. Setting `<connection-attempt-limit>999999</connection-attempt-limit>` will make all invocations hang trying (practically) indefinitely. 
2. Setting `<connection-attempt-limit>999999</connection-attempt-limit>` and  the system property `hazelcast.client.invocation.timeout.seconds=2` will work when reconnecting, but for the initial connection the client will hang indefinitely (until the server gets reachable). System property `hazelcast.client.invocation.timeout.seconds` does not seem to work when no connection has been ever made.
3. Defining `<connection-attempt-limit>999999</connection-attempt-limit>` and `<connection-strategy async-start="false" reconnect-mode="ASYNC">` will again work when reconnecting, but the client will hang indefinitely during the initial connection.
4. Defining `<connection-attempt-limit>999999</connection-attempt-limit>` and `<connection-strategy async-start="true" reconnect-mode="ASYNC">` will work on reconnect, but it has the unacceptable drawback that first invocations will (very likely) fail because the client didn't have enough time to connect. 

In alternatives 1, 2, and 3 (unlimited synchronous connection) our Spring application will get unresponsive without available threads because threads will wait for a Hazelcast connection.

**Other thoughts**
There is another workaround like throwing the current client and initializing a new one on error, but this has a relatively big performance penalty (the client is not lightweight at all with all its threads)  and this is not possible to implement in our case because it is a third-party library that controls the client.

I know the retry feature was implemented with the same idea, but it seems not enough and I consider this feature will make the Hazelcast connection more resilient for most applications .

**Comment by mmedenjak** [comment]
@entbugger thank you for the detailed report and suggestion! Handing off to @hazelcast/apis for consideration, please review and comment.

**Comment by ihsandemir** [comment]
Hello @entbugger The 1st alternative uses the default invocation timeout of 120 seconds, hence client invocation should be returning after 120secs. 

The 2nd alternative, should be working with 2 seconds invocation timeout. Can you please tell which API call are you talking about when you say `client.invoke()`?  Do you have a reproducer code for the problem? By the way, can you also share your client logs related to the problem?

**Comment by entbugger** [comment]
> The 1st alternative uses the default invocation timeout of 120 seconds, hence client invocation should be returning after 120secs.

The problem in the 1st, 2nd and 3rd alternatives is that the Java code hangs during the client construction (around  `HazelcastClient.newHazelcastClient(clientConfig);`). Strictly speaking from the Hazelcast client point of view it is not a client invocation because the client is not yet constructed, but from the view of our application  it is a client invocation.
The log looks like this: 

```
Sep. 07, 2021 7:45:10 PM com.hazelcast.client.connection.nio.ClusterConnectorService
INFO: hz.client_0 [bla] [3.12] Trying to connect to cluster with name: bla
Sep. 07, 2021 7:45:10 PM com.hazelcast.client.connection.nio.ClusterConnectorService
INFO: hz.client_0 [bla] [3.12] Unable to get live cluster connection, retry in 3000ms, attempt 1 of 999999.
Sep. 07, 2021 7:45:13 PM com.hazelcast.client.connection.nio.ClusterConnectorService
INFO: hz.client_0 [bla] [3.12] Unable to get live cluster connection, retry in 3000ms, attempt 2 of 999999.
Sep. 07, 2021 7:45:16 PM com.hazelcast.client.connection.nio.ClusterConnectorService
INFO: hz.client_0 [bla] [3.12] Unable to get live cluster connection, retry in 3000ms, attempt 3 of 999999.
...
Sep. 07, 2021 7:50:23 PM com.hazelcast.client.connection.nio.ClusterConnectorService
INFO: hz.client_0 [bla] [3.12] Unable to get live cluster connection, retry in 3000ms, attempt 101 of 999999.
...
```

> The 2nd alternative, should be working with 2 seconds invocation timeout. 
Yes, but during client construction it does not work. 

> Can you please tell which API call are you talking about when you say `client.invoke()`?
The exact Hazelcast API call is is hidden in the third-party library, but with `client.invoke()` I mean this (summed up): 
```
if (client==null) {
  client = HazelcastClient.newHazelcastClient(clientConfig);  
}
client.get("key");
```


> Do you have a reproducer code for the problem? By the way, can you also share your client logs related to the problem?
See the log above. The problem is easy to understand from the log above. I do not have reproducer code (do not have an IDE now), but I hope the problem is easy to reproduce when trying to connect during no server is listening.

**Comment by sancar** [comment]
Hi @entbugger 
I think you haven't tried the following configuration. Can you check it out ?

connection-attempt-limit>999999</connection-attempt-limit> and the system property hazelcast.client.invocation.timeout.seconds=2 
<connection-strategy async-start="true" reconnect-mode="ON">

 `async-start="true"` makes sure that you will not wait for initial setup to complete on newHazelcastClient
`reconnect-mode="ON"` makes sures that you will not get `HazelcastCLientOfflineException` immediately  in map.get/put operations when connections is not established yet. 
`hazelcast.client.invocation.timeout.seconds=2` makes sure that your map.get/put will not wait more than 2 seconds when there is a connection problem to the servers
`<connection-attempt-limit>999999</connection-attempt-limit>  ` and this one makes sure that the client will not shutdown for a long time.

**Comment by entbugger** [comment]
Hi @sancar
Thank you for your suggestion. I tried `<connection-attempt-limit>999999</connection-attempt-limit>` and `<connection-strategy async-start="true" reconnect-mode="ON">` and `hazelcast.client.invocation.timeout.seconds=2` with Hazelcast client 3.12 and 3.12.12. The result is that it still immediately throws  `com.hazelcast.client.HazelcastClientOfflineException: Client connecting to cluster`.

**Comment by sancar** [comment]
@entbugger Yes, you are right. I just tried that setup myself and my statement was wrong. 

It seems that we have made an exception for the initial setup where you can no longer control the behavior via `reconnect-mode`. and I understand that this makes your use-case problematic. And I think, your use-case is a very common one. 
I will discuss this within the team to see if how we can address it. Note that we no longer backport to 3.12.x. Even if we decide on a fix, it will be in 4.x series. 

Since you mentioned a workaround with throwing away the client and initializing new one, I would like to suggest another workaround. And I think, in the ideal situation, we should not make our users implement this and this should be a configuration.

```
    public static void main(String[] args) throws Exception {
        ClientConfig clientConfig = new ClientConfig();
        clientConfig.getConnectionStrategyConfig().setAsyncStart(true).setReconnectMode(ClientConnectionStrategyConfig.ReconnectMode.ON);
        getWithSingleTimeout(clientConfig, "key");
    }

    public static String getWithSingleTimeout(ClientConfig clientConfig, String key) throws InterruptedException {
        if (client == null) {
           //assuming there is single thread. Otherwise we need synchronization around this
            client = HazelcastClient.newHazelcastClient(clientConfig);
        }
        long beginTime = System.currentTimeMillis();
        while (System.currentTimeMillis() - beginTime < invocationTimeoutMillis) {
            try {
                return (String) client.getMap("map").get(key);
            } catch (HazelcastClientOfflineException e) {
                Thread.sleep(1000);
            }
        }
        throw new OperationTimeoutException();
    }
  ```

I would like leave this issue open until a decision made, and thanks again for the feedback @entbugger .  I think, this is a valuable one.
